### PR TITLE
Use Opacus 1.13 instead of 0.15

### PR DESCRIPTION
**Pull request by amanoel**

### CODE REVIEW

File: src/dp_transformers/grad_sample/transformers/tied_embedding.py [hunk 1]
```diff
@@ -38,8 +40,8 @@ def compute_tied_embedding_grad_sample(
     grad_sample.scatter_add_(1, index, B.reshape(batch_size, -1, layer.embedding_dim))
     torch.backends.cudnn.deterministic = saved
 
-    utils.create_or_accumulate_grad_sample(layer.weight, grad_sample, layer)
```
Comment (huseyinatahaninan) [inline src/dp_transformers/grad_sample/transformers/tied_embedding.py:41]: Before Opacus 1.0 they had this function `create_or_accumulate_grad_sample` to add two or more gradients coming from tied parameters. New version removed this function but then they fixed it with some other approach https://github.com/pytorch/opacus/issues/413. This change would not add up to the gradients but actually replace it so it would not be right. We can remove this script altogether as it's being handled automatically now afaik :)

File: src/dp_transformers/grad_sample/transformers/conv_1d.py [hunk 2]
```diff
@@ -1,27 +1,22 @@
...
-        utils.create_or_extend_grad_sample(
-            layer.bias,
-            torch.einsum("n...k->nk", B),
-            batch_dim,  # pyre-ignore[6] We know layer.bias is not None
-        )
-
+        ret[layer.bias] = torch.einsum("n...k->nk", B)
+    return ret
 
 def register_grad_sampler() -> None:
```
Comment (huseyinatahaninan) [inline src/dp_transformers/grad_sample/transformers/conv_1d.py:21]: Can we remove this and do it like the new version does with the decorators? https://github.com/pytorch/opacus/blob/main/opacus/grad_sample/embedding.py

I think then we'd just need to import the file instead of calling this function

File: src/dp_transformers/dp_utils.py [hunk 3]
```diff
@@ -59,14 +63,20 @@ def on_step_end(self, args: training_args.TrainingArguments, state: TrainerState
                 "which has an appropriate callback in the trainer."
             )
 
+        if optimizer is None:
+            raise RuntimeError("Impossible to access optimizer from inside callback")
+        optimizer.signal_skip_step(do_skip=False)
```
Comment (huseyinatahaninan) [inline src/dp_transformers/dp_utils.py:68]: Should this not be on_step_begin instead?

File: src/dp_transformers/dp_utils.py [hunk 4]
```diff
@@ -174,14 +136,93 @@ class OpacusDPTrainer(Trainer):
...
+        # Set up callback for accounting and handling grad acc
+        self.dp_callback = DPCallback(
+            noise_multiplier=self.privacy_args.noise_multiplier,
+            target_delta=self.privacy_args.target_delta,
+            sampling_probability=self.sampling_probability,
+            compute_epsilon=accountant_fn
+        )
+        super().__init__(model=model, args=args, train_dataset=train_dataset, callbacks=[self.dp_callback], **kwargs)
+
+        self.get_rdp_epsilon = lambda: self.dp_callback.accountant.get_epsilon(self.privacy_args.target_delta)  # RDP epsilon
```
Comment (wulu473) [inline src/dp_transformers/dp_utils.py:197]: This seems a bit circular and not easy to follow. `Trainer.privacy_accountant` is a PRV accountant but `DPCallback.accountant` is a RDP accountant. Should we try to make the trainer have all the accountants and give them more interpretable names?

File: src/dp_transformers/dp_utils.py [hunk 5]
```diff
@@ -174,14 +136,93 @@ class OpacusDPTrainer(Trainer):
...
+    @property
+    def sampling_probability(self) -> float:
+        return self.train_args.per_device_train_batch_size * self.train_args.world_size * \
+            self.train_args.gradient_accumulation_steps / len(self.author_mapping)
+
+    @property
+    def num_steps(self) -> int:
+        return int(self.train_args.num_train_epochs * (1 / self.sampling_probability + 1))
+
+    def create_optimizer(self):
```
Comment (wulu473) [inline src/dp_transformers/dp_utils.py:209]: [I can't put a comment at the appropriate line]. Shouldn't we use the `DPDataLoader` for the `get_train_dataloader` function

Comment (huseyinatahaninan) [reply src/dp_transformers/dp_utils.py:209]: We don't really use `DPDataLoader` in our framework actually.

File: src/dp_transformers/dp_utils.py [hunk 6]
```diff
@@ -174,14 +138,89 @@ class OpacusDPTrainer(Trainer):
...
+            noise_multiplier=self.privacy_args.noise_multiplier,
+            target_delta=self.privacy_args.target_delta,
+            sampling_probability=self.sampling_probability,
+            rdp_accountant=self.rdp_accountant,
+            prv_accountant=self.prv_accountant
+        )
+        super().__init__(model=model, args=args, train_dataset=train_dataset, callbacks=[self.dp_callback], **kwargs)
+
+        self.get_rdp_epsilon = lambda: self.rdp_accountant.get_epsilon(self.privacy_args.target_delta)  # RDP epsilon
+        self.get_accountant_epsilon = lambda: self.prv_accountant.compute_epsilon(self.state.global_step)[2]
```
Comment (wulu473) [inline src/dp_transformers/dp_utils.py:196]: Should we also change this to `get_prv_epsilon`. I think this would make it a bit clearer.

File: src/dp_transformers/dp_utils.py (hunk 5 above)
Comment (wulu473) [reply src/dp_transformers/dp_utils.py:207]: I see. I think we should also use that since it's doing the correct poisson sampling. But we can do this in a follow up PR.

File: src/dp_transformers/dp_utils.py (hunk 6 above)
Comment (huseyinatahaninan) [reply src/dp_transformers/dp_utils.py]: certainly

**Review by huseyinatahaninan — APPROVED** [review]
